### ISSUE TITLE
[DTensor] DTensor is not well supported on older versions of GPUs, such as A10

### ISSUE THREAD

**Issue by Accelerator1996** [pytorch/pytorch#155657]
### 🐛 Describe the bug

When I run the test case **distributed/tensor/test_matrix_ops.py::DistMatrixOpsTest::test_dtensor_mm** on A10 * 2 machine, it failed. So I try to write a demo to run on other machines. The code is as below:
```
import torch
from torch.distributed._tensor.placement_types import Replicate, Shard
from torch.distributed.tensor import (
    DeviceMesh,
    distribute_module,
    distribute_tensor,
    DTensor,
    init_device_mesh,
)
import torch.distributed as dist

def dprint(*args):
    if dist.get_rank() == 0:
        print(*args)

mesh = init_device_mesh('cuda', (1, 1))
lhs = torch.randn(256, 128)
rhs = torch.randn(128, 256)
dprint(lhs)
dprint(rhs)
mm_result = lhs @ rhs
dprint(mm_result)

lhs_dtensor = distribute_tensor(lhs, mesh, [Shard(dim=0), Replicate()])
rhs_dtensor = distribute_tensor(rhs, mesh, [Replicate(), Shard(dim=1)])
dprint(lhs_dtensor)
dprint(lhs_dtensor.shape)
dprint(rhs_dtensor)
dprint(rhs_dtensor.shape)
dtensor_result = lhs_dtensor @ rhs_dtensor
dprint(dtensor_result)
```
Execute command:
```
torchrun --nproc-per-node=2 --master-port=22222 test_torch.py
```

Execute the code on H20 machine, we can see dtensor is the same as local tenser
```
tensor([[-1.2031,  0.7327, -1.4305,  ...,  1.4086,  1.4120,  0.3787],
        [-0.4451,  1.0876,  0.4238,  ..., -0.4681, -0.7065,  2.4794],
        [-1.8110, -0.7831,  0.1394,  ..., -0.9431, -0.1475, -1.0138],
        ...,
        [-0.3227,  0.5323,  1.6431,  ...,  0.8321,  0.0960, -0.3861],
        [ 0.4607, -0.0444,  0.6062,  ...,  0.7214, -0.2810, -0.5923],
        [-0.0457,  0.3255,  0.1792,  ...,  1.0951,  1.2481, -0.9751]])
tensor([[-0.4750,  0.3389, -1.3343,  ..., -0.5613, -0.3125, -2.2852],
        [ 0.7892,  0.2136,  0.1375,  ...,  1.3518,  0.8370, -0.5381],
        [ 0.5011, -0.4176,  0.3807,  ...,  0.1806, -0.6263,  1.8961],
        ...,
        [-0.3034,  1.2439, -0.4308,  ...,  0.3179, -1.6328, -1.1872],
        [-0.9453,  1.1993,  0.0509,  ...,  1.0084, -0.2565,  0.1930],
        [ 0.0943,  1.0572, -0.6163,  ...,  0.2439,  0.5112, -0.4624]])
tensor([[ 10.3300,  20.2515,   3.0012,  ...,  -3.4316,   2.7046,  18.1858],
        [-20.4380,   6.3968,  -7.4297,  ..., -10.4693, -24.5430,  -5.5365],
        [ -9.9097,   1.0555,  12.0763,  ...,  11.2950,  -8.5258,  12.3995],
        ...,
        [ 10.3550,  -2.0498,   6.6208,  ...,   5.2652, -18.5190,   7.9144],
        [ -1.0955,   0.8179, -16.1898,  ...,   8.5519,   1.1573,  18.2833],
        [  1.3768,  -3.3506,  -5.6115,  ...,   0.2833,   0.5255,   0.3142]])
DTensor(local_tensor=tensor([[-1.2031,  0.7327, -1.4305,  ...,  1.4086,  1.4120,  0.3787],
        [-0.4451,  1.0876,  0.4238,  ..., -0.4681, -0.7065,  2.4794],
        [-1.8110, -0.7831,  0.1394,  ..., -0.9431, -0.1475, -1.0138],
        ...,
        [-0.3227,  0.5323,  1.6431,  ...,  0.8321,  0.0960, -0.3861],
        [ 0.4607, -0.0444,  0.6062,  ...,  0.7214, -0.2810, -0.5923],
        [-0.0457,  0.3255,  0.1792,  ...,  1.0951,  1.2481, -0.9751]],
       device='cuda:0'), device_mesh=DeviceMesh('cuda', [[0]]), placements=(Shard(dim=0), Replicate()))
torch.Size([256, 128])
DTensor(local_tensor=tensor([[-0.4750,  0.3389, -1.3343,  ..., -0.5613, -0.3125, -2.2852],
        [ 0.7892,  0.2136,  0.1375,  ...,  1.3518,  0.8370, -0.5381],
        [ 0.5011, -0.4176,  0.3807,  ...,  0.1806, -0.6263,  1.8961],
        ...,
        [-0.3034,  1.2439, -0.4308,  ...,  0.3179, -1.6328, -1.1872],
        [-0.9453,  1.1993,  0.0509,  ...,  1.0084, -0.2565,  0.1930],
        [ 0.0943,  1.0572, -0.6163,  ...,  0.2439,  0.5112, -0.4624]],
       device='cuda:0'), device_mesh=DeviceMesh('cuda', [[0]]), placements=(Replicate(), Shard(dim=1)))
torch.Size([128, 256])
DTensor(local_tensor=tensor([[ 10.3300,  20.2515,   3.0012,  ...,  -3.4316,   2.7046,  18.1858],
        [-20.4380,   6.3968,  -7.4297,  ..., -10.4693, -24.5430,  -5.5365],
        [ -9.9097,   1.0555,  12.0763,  ...,  11.2950,  -8.5258,  12.3995],
        ...,
        [ 10.3550,  -2.0498,   6.6208,  ...,   5.2652, -18.5190,   7.9144],
        [ -1.0955,   0.8179, -16.1898,  ...,   8.5519,   1.1573,  18.2833],
        [  1.3768,  -3.3506,  -5.6115,  ...,   0.2833,   0.5255,   0.3142]],
       device='cuda:0'), device_mesh=DeviceMesh('cuda', [[0]]), placements=(Shard(dim=0), Shard(dim=1)))
```
Execute the code on A10 machine, we can see mm result is different
```
tensor([[ 0.6455, -0.8630, -0.2597,  ..., -2.5630, -0.0319,  2.0597],
        [-1.1884, -0.3261, -0.6148,  ...,  0.1014, -0.0036, -1.0843],
        [ 0.2959, -0.4018, -0.9721,  ...,  0.5594,  0.8147, -0.6585],
        ...,
        [ 0.5468, -2.2421,  1.5114,  ..., -1.1265,  1.6534,  0.5183],
        [ 0.8433,  1.1677, -0.9121,  ...,  1.5670, -1.6668,  0.7243],
        [ 0.0460,  0.0643, -0.2050,  ...,  0.6881, -0.2633,  0.5044]])
tensor([[-1.7412,  0.7234,  1.7052,  ..., -1.9299, -0.9990, -0.5263],
        [ 1.0690, -0.6155,  0.6466,  ..., -0.5912,  1.3025, -1.2864],
        [ 0.8885,  0.5027,  1.0266,  ..., -1.5499, -1.2875,  1.4335],
        ...,
        [ 0.1859, -1.4796,  0.0859,  ...,  0.6313, -1.0837,  0.4538],
        [-0.0303, -0.1001, -0.4886,  ..., -0.0542,  1.6519,  1.0781],
        [ 0.1298,  0.1236,  0.2947,  ...,  0.2576,  0.5009,  1.2897]])
tensor([[  2.7477,   9.9757,  -6.3250,  ...,   0.5026,  11.5428,  -1.9447],
        [ -2.1091,   6.5683, -18.9510,  ...,   8.0808,  -4.0913,   4.4008],
        [ 13.1233,   5.3387,   3.9201,  ...,  20.1069,   6.6060,  -4.1678],
        ...,
        [ -8.9132,  17.7708,   4.6434,  ...,  16.0531,  -2.2469,  15.8246],
        [ 13.5751,   1.0290,  11.5692,  ...,   9.0137, -20.7020,  -6.5653],
        [  2.5063,  -0.9000,   2.2846,  ...,  22.2670,   0.3579,  -8.0632]])
DTensor(local_tensor=tensor([[ 0.6455, -0.8630, -0.2597,  ..., -2.5630, -0.0319,  2.0597],
        [-1.1884, -0.3261, -0.6148,  ...,  0.1014, -0.0036, -1.0843],
        [ 0.2959, -0.4018, -0.9721,  ...,  0.5594,  0.8147, -0.6585],
        ...,
        [ 0.5468, -2.2421,  1.5114,  ..., -1.1265,  1.6534,  0.5183],
        [ 0.8433,  1.1677, -0.9121,  ...,  1.5670, -1.6668,  0.7243],
        [ 0.0460,  0.0643, -0.2050,  ...,  0.6881, -0.2633,  0.5044]],
       device='cuda:0'), device_mesh=DeviceMesh('cuda', [[0]]), placements=(Shard(dim=0), Replicate()))
torch.Size([256, 128])
DTensor(local_tensor=tensor([[-1.7412,  0.7234,  1.7052,  ..., -1.9299, -0.9990, -0.5263],
        [ 1.0690, -0.6155,  0.6466,  ..., -0.5912,  1.3025, -1.2864],
        [ 0.8885,  0.5027,  1.0266,  ..., -1.5499, -1.2875,  1.4335],
        ...,
        [ 0.1859, -1.4796,  0.0859,  ...,  0.6313, -1.0837,  0.4538],
        [-0.0303, -0.1001, -0.4886,  ..., -0.0542,  1.6519,  1.0781],
        [ 0.1298,  0.1236,  0.2947,  ...,  0.2576,  0.5009,  1.2897]],
       device='cuda:0'), device_mesh=DeviceMesh('cuda', [[0]]), placements=(Replicate(), Shard(dim=1)))
torch.Size([128, 256])
DTensor(local_tensor=tensor([[  2.7530,   9.9736,  -6.3227,  ...,   0.5043,  11.5500,  -1.9456],
        [ -2.1154,   6.5692, -18.9514,  ...,   8.0787,  -4.0936,   4.4053],
        [ 13.1229,   5.3360,   3.9216,  ...,  20.1030,   6.6072,  -4.1663],
        ...,
        [ -8.9133,  17.7660,   4.6387,  ...,  16.0515,  -2.2449,  15.8283],
        [ 13.5778,   1.0275,  11.5734,  ...,   9.0167, -20.7056,  -6.5633],
        [  2.5072,  -0.8948,   2.2881,  ...,  22.2679,   0.3575,  -8.0691]],
       device='cuda:0'), device_mesh=DeviceMesh('cuda', [[0]]), placements=(Shard(dim=0), Shard(dim=1)))
```

### Versions

docker image: nvcr.io/nvidia/pytorch:25.05-py3
torch version: 2.7.1 # I download it by piping install to replace 2.7.0 in image

**Comment by weifengpy** [comment]
cc @XilunWu for DTensor on A10 gpus

**Comment by XilunWu** [comment]
This could be a bug.

For this `(Shard(dim=0), Replicate())` @ `(Replicate(), Shard(dim=1))` matmul case, there should be no data redistribution but pure local matmul. Since I don't have an A10 GPU to reproduce, what I can help is to provide a modified code based on your example so that you can run it and share us the output. I'll reply with the code soony.
